### PR TITLE
Stop creating unnecessary overdue charge when amount is 0.

### PR DESCRIPTION
**Pull request by mohitsinha**
## Description
Describe the changes made and why they were made. Ignore if these details are present on the associated Jira ticket

## Checklist
Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Commit message starts with the issue number from https://issues.apache.org/jira/projects/FINERACT/. Ex: FINERACT-646 Pockets API.

- [x] Coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions have been followed.

- [x] API documentation at https://github.com/apache/fineract/blob/develop/api-docs/apiLive.htm has been updated with details of any API changes.

- [x] Integration tests have been created/updated for verifying the changes made.

- [x] All Integrations tests are passing with the new commits.

- [x] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the list.)

Our guidelines for code reviews is at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide

### CODE REVIEW

**Comment by mohitsinha** [comment]
@conradsp @vishwasbabu @awasum Can anyone review this?

**Comment by mohitsinha** [comment]
@vorburger

**Comment by cnangeh** [comment]
@mohitsinha the code update looks good. Can you update the necessary unit test to reflect this new update?

**Comment by mohitsinha** [comment]
@Anh3h Hey, thanks for the review. I didn't find any test for existing crons. So, it will be difficult for me to setup everything from scratch. If it's any helpful, I have done manual testing for this.

**Comment by mohitsinha** [comment]
@vorburger Can this be merged without a test case?

**Comment by awasum** [comment]
@mohitsinha It will be helpful to take sometime and add tests for the feature you provided. If there is no test for that module, then its recommended to invest some time and put tests for maintenance purposes.

Imagine the next person who works on this particular file or feature type also does not write a test, it means over time, we will keep merging without tests. This can create a situation where we are creating features without test all because.. "it works manually".  Set the foundation so your future should be easy to manage.

**Comment by mohitsinha** [comment]
@Anh3h The tests are added.

**Comment by cnangeh** [comment]
This looks to me.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java
Comment (vorburger) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java]: is there really a realistic case when `overdueDTO` can be `null` ? If not, then this could hide future bugs, and it would be better to remove this and let it explode with a NullPointerException to find such bugs, instead of silently ignoring a condition which shold never happen. If it can happen, then it's fine - but I don't see how.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java
Comment (vorburger) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java]: This would be also true not only if `loanCharge.amount()` is 0 but also if it's negative. Is that really intended?

File: fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/SchedulerJobsTestResults.java
Comment (vorburger) [inline fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/SchedulerJobsTestResults.java:728]: I'm unable to say if this is functionally correct - would someone else be able to review and confirm it is?

Comment (vorburger) [reply fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/SchedulerJobsTestResults.java:728]: @awasum @Anh3h do you understand more about the functionality here than I do and can you confirm this new test is functionally correct?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java
Comment (mohitsinha) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java]: Not really. I did it just to be safe.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java
Comment (mohitsinha) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java]: ```applyChargeToOverdueLoanInstallment``` won't ever return a null `overdueDTO` as the return itself is a constructor call. I can let it explode but the problem is this whole logic is inside a for loop for all loans that are overdue. Hence, if this explodes it will impact the normal flow for overdue penalty for other loans. Hence it had to be handled gracefully. I can let it explode with a `try catch block`

File: fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/SchedulerJobsTestResults.java
Comment (mohitsinha) [reply fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/SchedulerJobsTestResults.java:728]: I can throw some light into this. 
It is creating a loan with a very small interest rate 0.000001%.
It is verifying that the `totalChargesDue` is 0 for the loan ( which is also the feature currently as 0+ 0 + 0 is 0).
I am then calling the api `/loans/{id}?associations=charges` and then verifying that there are no charges created for this loan. This actually tests correctly what I have done.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java
Comment (vorburger) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java]: > `applyChargeToOverdueLoanInstallment` won't ever return a null `overdueDTO` as the return itself is a constructor call. 

OK but then you agree that this null check is unnecessary? If as you say it's impossible for `overdueDTO` to ever be null (because `applyChargeToOverdueLoanInstallment` ends with a `return new LoanOverdueDTO`) then why add this nul lcheck as part of this PR?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java
Comment (vorburger) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java]: would it be clearer for future readers of this code to check only for `equals(ZERO)` instead? Just a thought.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java
Comment (mohitsinha) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java]: `applyChargeToOverdueLoanInstallment` is  creating a `LoanCharge` and returning `LoanOverdueDTO`. I want to avoid creating the `LoanOverdueInstallmentCharge` on certain conditions ( the amount is zero). As the function holds all the logic for loan charges and returns using a constructor, I had to `return null` to escape the logic of charge creation. 

Hence the null pointer check is due to my code changes inside `applyChargeToOverdueLoanInstallment`

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java
Comment (mohitsinha) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java]: Sure. I will make the change

Comment (mohitsinha) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java]: `equals` won't work because it also takes into consideration the scale. I have changed it to use `compareTo == 0`

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java
Comment (vorburger) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java]: oh right, now I get it. Sorry, I was a bit slow on this one! :smile_cat:

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java
Comment (vorburger) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java]: OK, cool - resolves conversation for me; thanks!

**Comment by vorburger** [comment]
@awasum @Anh3h this LGTM now, and I have no more objections. I'm hesitant to merge it myself, as I suspect that you understand more about the functionality here than I do and can you confirm that the implementation is fine and the new test is functionally correct... could I let you review and merge this? Tx.

**Review by vorburger — APPROVED** [review]

**Comment by mohitsinha** [comment]
@awasum @Anh3h Do you feel anything is still pending on this one?

**Comment by awasum** [comment]
I dont really understand this part of the software very well. I feel we can just merge and if we get any problems, we revert. 
For now, Please @vishwasbabu  will you have time to review this within the next few days after which you can merge or I will just do that after 48 hours.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java
Comment (angelboxes) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java]: I don't think its okay to return null in this line, as you can see in this part of the code there is a for statement that iterates through a set of dates. If the amount for the charge in the first date entry were to be 0, the next dates wouldn't create any new charges even if the charge amount is suposed to be greater than 0.

I suggest to replace the return null statement with a continue statement, to continue iterating the entry set

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java
Comment (angelboxes) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java]: I made another comment why null shouldn't be returned by the function applyChargeToOverdueLoanInstallment, so I suggest to also remove this if statement.

File: fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/SchedulerJobsTestResults.java
Comment (angelboxes) [inline fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/SchedulerJobsTestResults.java:762]: Executing the "Apply penalties" job doesn't guarantee you that the created loan is processed by the time this test finishes its run. It will only start the job. Before checking the loan charges, you should check either if the job finished or if the loan was processed. This job will also check penalties for any other active loan with charges and depending on loan configuration and  the date it was created, this could take some time.

**Review by angelboxes — CHANGES_REQUESTED** [review]
I added some comments about the changes made in the applyChargeToOverdueLoanInstallment function that should be addressed just in case.

File: fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/SchedulerJobsTestResults.java
Comment (mohitsinha) [reply fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/SchedulerJobsTestResults.java:762]: Yes, I totally agree with this. The Job usually takes some time when the number of loans have increased.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java
Comment (mohitsinha) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java]: Currently we have only these types of `ChargeCalculationType`.
```    FLAT(1, "chargeCalculationType.flat"), //
    PERCENT_OF_AMOUNT(2, "chargeCalculationType.percent.of.amount"), //
    PERCENT_OF_AMOUNT_AND_INTEREST(3, "chargeCalculationType.percent.of.amount.and.interest"), //
    PERCENT_OF_INTEREST(4, "chargeCalculationType.percent.of.interest"),
    PERCENT_OF_DISBURSEMENT_AMOUNT(5,"chargeCalculationType.percent.of.disbursement.amount");
```
Even though it seems, that the amount can be different for a different iteration, in fact it is always the same.

File: fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/SchedulerJobsTestResults.java
Comment (mohitsinha) [reply fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/SchedulerJobsTestResults.java:762]: In fact, this check should be there in all the tests which rely on jobs to be completed.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java
Comment (angelboxes) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java]: Ok that may be right I'm just saying it as I don't take anything for granted and also it will avoid returning a null value as the overdueDTO.

Also, something similiar it could happen in the applyOverdueChargesForLoan function with the if statement:

if(overdueDTO == null){
     return;
}

Because you are iterating a list of multiple overdue installments for the same loan. If the penalty charge is calculated based on the installment interest amount, that amount can be different for each one of those installments, for the first one the calculated amount for all the installment charges can be 0, but for the next ones it could be a value greater than zero. Making a return call would not calculate charges for the next overdue installments.

File: fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/SchedulerJobsTestResults.java
Comment (angelboxes) [reply fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/SchedulerJobsTestResults.java:762]: > In fact, this check should be there in all the tests which rely on jobs to be completed.

I agree,, in this case this the only way that loan penalty charges can be created,, but the job running time is something we should be concerned, by personal experience I can tell you that a loan penalty charge with daily  frequency could take forever specially if the overdue date is very old.

Comment (mohitsinha) [reply fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/SchedulerJobsTestResults.java:762]: True. I made a change for that.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java
Comment (mohitsinha) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java]: I have made those changes.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java
Comment (mohitsinha) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java]: I have made those changes.

File: fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/SchedulerJobsTestResults.java
Comment (angelboxes) [reply fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/SchedulerJobsTestResults.java:762]: Ok, it seems right.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java
Comment (angelboxes) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java]: Ok, it seems right. 
You can be right about charge calculations as actually I'm not too concerned about how exactly charges amount are calculated if they always return the same amount, but skipping every other loan overdue installments could also ignore some that may have a different charge definition and this needed to be addressed at least in this function. Thanks.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java
Comment (angelboxes) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java]: It seems right. Thanks

**Review by angelboxes — APPROVED** [review]

**Comment by vorburger** [comment]
@angelboxes sorry I forgot if you are a committer? If yes, please do go ahead and merge PRs, such as this, and others, which LGTY... you don't have to wait e.g. for me; on the contrary, I very much hope for more active other committers who review and merge.

**Comment by awasum** [comment]
Go for it @angelboxes .

**Comment by angelboxes** [comment]
> @angelboxes sorry I forgot if you are a committer? If yes, please do go ahead and merge PRs, such as this, and others, which LGTY... you don't have to wait e.g. for me; on the contrary, I very much hope for more active other committers who review and merge.

Yes I'm a committer. 

Wow!, I feel hesitant as I always like to see if there's something left to be said by someone else, But if I have the go-ahead by you I will merge it right now.

**Comment by awasum** [comment]
There is always revert...

**Pull request merged**
